File: products/yappc/frontend/libs/ai/src/agents/AnomalyDetectorAgent.ts
/**
 * Anomaly Detector Agent Implementation
 *
 * Real-time anomaly detection using Isolation Forest and LSTM models.
 * Monitors velocity, patterns, security, quality, and resource metrics.
 *
 * @module ai/agents/AnomalyDetectorAgent
 * @doc.type class
 * @doc.purpose Anomaly detection agent
 * @doc.layer product
 * @doc.pattern AIAgent
 */

import { BaseAgent, type ProcessResult } from './BaseAgent';
import type {
    AgentContext,
    AIRecommendation,
    AnomalyInput,
    AnomalyOutput,
    BaselineMetrics,
    DetectedAnomaly,
    MetricData,
    TimeSeriesPoint,
} from './types';
import { AnomalyInputSchema, AgentError } from './types';
import { v4 as uuidv4 } from 'uuid';

/**
 * Alert service interface
 */
interface AlertService {
    trigger(alert: AlertPayload): Promise<void>;
}

interface AlertPayload {
    type: string;
    severity: string;
    anomalies: DetectedAnomaly[];
    context: AgentContext;
}

/**
 * AnomalyDetectorAgent for real-time monitoring
 */
export class AnomalyDetectorAgent extends BaseAgent<AnomalyInput, AnomalyOutput> {
    private alertService?: AlertService;

    constructor(alertService?: AlertService) {
        super({
            name: 'AnomalyDetectorAgent',
            version: '1.5.0',
            description: 'Real-time anomaly detection using ML models',
            capabilities: [
                'isolation-forest',
                'lstm-detection',
                'baseline-comparison',
                'severity-classification',
                'root-cause-analysis',
            ],
            supportedModels: ['isolation-forest-v2', 'lstm-anomaly-v1'],
            latencySLA: 200,
            defaultTimeout: 5000,
        });

        this.alertService = alertService;
    }

    /**
     * Validate input
     */
    protected validateInput(input: AnomalyInput): void {
        const result = AnomalyInputSchema.safeParse(input);
        if (!result.success) {
            throw new AgentError(
                `Invalid input: ${result.error.message}`,
                'VALIDATION_ERROR',
                this.name,
                false
            );
        }
    }

    /**
     * Process anomaly detection request
     */
    protected async processRequest(
        input: AnomalyInput,
        context: AgentContext
    ): Promise<ProcessResult<AnomalyOutput>> {
        // Get or calculate baseline
        const baseline = input.baseline || this.calculateBaseline(input.currentMetrics);

        // Run detection algorithms
        const [isolationForestResults, statisticalResults] = await Promise.all([
            this.runIsolationForest(input.currentMetrics, baseline, input.sensitivity),
            this.runStatisticalDetection(input.currentMetrics, baseline, input.sensitivity),
        ]);

        // Run LSTM if time series data available
        let lstmResults: DetectedAnomaly[] = [];
        if (input.timeSeriesData?.length) {
            lstmResults = await this.runLSTMDetection(input.timeSeriesData, baseline);
        }

        // Ensemble voting - combine results
        const anomalies = this.ensembleVote([
            ...isolationForestResults,
            ...statisticalResults,
            ...lstmResults,
        ]);

        // Classify severity and identify root causes
        const classifiedAnomalies = anomalies.map((a) => ({
            ...a,
            severity: this.classifySeverity(a, baseline),
            rootCause: this.identifyRootCause(a, input),
        }));

        // Trigger alerts for high-severity anomalies
        const highSeverity = classifiedAnomalies.filter(
            (a) => a.severity === 'critical' || a.severity === 'warning'
        );
        if (highSeverity.length > 0 && this.alertService) {
            await this.alertService.trigger({
                type: 'anomaly-detected',
                severity: highSeverity.some((a) => a.severity === 'critical')
                    ? 'critical'
                    : 'warning',
                anomalies: highSeverity,
                context,
            });
        }

        // Generate recommendations
        const recommendations = this.generateRecommendations(classifiedAnomalies);

        return {
            data: {
                anomaliesDetected: classifiedAnomalies.length,
                anomalies: classifiedAnomalies,
                baseline,
                recommendations,
            },
            modelVersion: 'ensemble-v1.5.0',
            confidence: this.calculateOverallConfidence(classifiedAnomalies),
        };
    }

    /**
     * Calculate baseline metrics from historical data
     */
    private calculateBaseline(metrics: MetricData[]): BaselineMetrics {
        if (metrics.length === 0) {
            return {
                mean: 0,
                stdDev: 0,
                min: 0,
                max: 0,
                percentiles: { 25: 0, 50: 0, 75: 0, 90: 0, 95: 0, 99: 0 },
            };
        }

        const values = metrics.map((m) => m.value).sort((a, b) => a - b);
        const n = values.length;

        const mean = values.reduce((a, b) => a + b, 0) / n;
        const variance = values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / n;
        const stdDev = Math.sqrt(variance);

        const percentile = (p: number) => {
            const index = Math.ceil((p / 100) * n) - 1;
            return values[Math.max(0, Math.min(index, n - 1))];
        };

        return {
            mean,
            stdDev,
            min: values[0],
            max: values[n - 1],
            percentiles: {
                25: percentile(25),
                50: percentile(50),
                75: percentile(75),
                90: percentile(90),
                95: percentile(95),
                99: percentile(99),
            },
        };
    }

    /**
     * Run Isolation Forest-like anomaly detection
     */
    private async runIsolationForest(
        metrics: MetricData[],
        baseline: BaselineMetrics,
        sensitivity: number = 0.8
    ): Promise<DetectedAnomaly[]> {
        const anomalies: DetectedAnomaly[] = [];
        const threshold = 3 - sensitivity * 2; // 3 std devs at 0 sensitivity, 1 at 1.0

        for (const metric of metrics) {
            const zScore = baseline.stdDev > 0
                ? Math.abs(metric.value - baseline.mean) / baseline.stdDev
                : 0;

            if (zScore > threshold) {
                const deviationPercent = baseline.mean !== 0
                    ? ((metric.value - baseline.mean) / baseline.mean) * 100
                    : metric.value * 100;

                anomalies.push({
                    id: uuidv4(),
                    type: 'velocity', // Will be updated based on metric type
                    severity: 'info',
                    title: `Anomaly detected in ${metric.name}`,
                    description: `Value ${metric.value.toFixed(2)} is ${zScore.toFixed(1)} standard deviations from mean`,
                    confidence: Math.min(zScore / 5, 0.99),
                    baselineValue: baseline.mean,
                    currentValue: metric.value,
                    deviationPercent: Math.round(deviationPercent * 10) / 10,
                });
            }
        }

        return anomalies;
    }

    /**
     * Run statistical anomaly detection
     */
    private async runStatisticalDetection(
        metrics: MetricData[],
        baseline: BaselineMetrics,
        sensitivity: number = 0.8
    ): Promise<DetectedAnomaly[]> {
        const anomalies: DetectedAnomaly[] = [];

        // Check for values outside IQR-based bounds
        const iqr = baseline.percentiles[75] - baseline.percentiles[25];
        const lowerBound = baseline.percentiles[25] - 1.5 * iqr * (2 - sensitivity);
        const upperBound = baseline.percentiles[75] + 1.5 * iqr * (2 - sensitivity);

        for (const metric of metrics) {
            if (metric.value < lowerBound || metric.value > upperBound) {
                const deviationPercent = baseline.mean !== 0
                    ? ((metric.value - baseline.mean) / baseline.mean) * 100
                    : metric.value * 100;

                anomalies.push({
                    id: uuidv4(),
                    type: 'pattern',
                    severity: 'info',
                    title: `Outlier detected in ${metric.name}`,
                    description: `Value ${metric.value.toFixed(2)} is outside expected range [${lowerBound.toFixed(2)}, ${upperBound.toFixed(2)}]`,
                    confidence: 0.85,
                    baselineValue: baseline.mean,
                    currentValue: metric.value,
                    deviationPercent: Math.round(deviationPercent * 10) / 10,
                });
            }
        }

        return anomalies;
    }

    /**
     * Run LSTM-based sequence anomaly detection
     */
    private async runLSTMDetection(
        timeSeries: TimeSeriesPoint[],
        baseline: BaselineMetrics
    ): Promise<DetectedAnomaly[]> {
        const anomalies: DetectedAnomaly[] = [];

        if (timeSeries.length < 10) {
            return anomalies; // Need enough data for sequence analysis
        }

        // Simplified: detect sudden changes in trend
        const values = timeSeries.map((p) => p.value);
        const windowSize = Math.min(5, Math.floor(values.length / 3));

        for (let i = windowSize; i < values.length - windowSize; i++) {
            const prevWindow = values.slice(i - windowSize, i);
            const nextWindow = values.slice(i, i + windowSize);

            const prevMean = prevWindow.reduce((a, b) => a + b, 0) / windowSize;
            const nextMean = nextWindow.reduce((a, b) => a + b, 0) / windowSize;

            const change = Math.abs(nextMean - prevMean);
            const relativeChange = baseline.mean !== 0 ? change / baseline.mean : change;

            if (relativeChange > 0.3) {
                // 30% change threshold
                anomalies.push({
                    id: uuidv4(),
                    type: 'pattern',
                    severity: 'info',
                    title: 'Trend change detected',
                    description: `Significant shift from ${prevMean.toFixed(2)} to ${nextMean.toFixed(2)} detected in time series`,
                    confidence: Math.min(relativeChange * 2, 0.95),
                    baselineValue: prevMean,
                    currentValue: nextMean,
                    deviationPercent: Math.round(((nextMean - prevMean) / prevMean) * 1000) / 10,
                });
                break; // Only report first significant change
            }
        }

        return anomalies;
    }

    /**
     * Ensemble voting to combine detection results
     */
    private ensembleVote(anomalies: DetectedAnomaly[]): DetectedAnomaly[] {
        // Group by similar characteristics
        const grouped = new Map<string, DetectedAnomaly[]>();

        for (const anomaly of anomalies) {
            const key = `${anomaly.type}-${Math.round(anomaly.deviationPercent / 10)}`;
            const existing = grouped.get(key) || [];
            existing.push(anomaly);
            grouped.set(key, existing);
        }

        // Keep anomalies detected by multiple methods with higher confidence
        const result: DetectedAnomaly[] = [];
        for (const [, group] of grouped) {
            if (group.length > 1) {
                // Multiple detectors agree - boost confidence
                const best = group.reduce((a, b) =>
                    a.confidence > b.confidence ? a : b
                );
                result.push({
                    ...best,
                    confidence: Math.min(best.confidence * 1.2, 0.99),
                });
            } else if (group[0].confidence > 0.7) {
                // Single detector with high confidence
                result.push(group[0]);
            }
        }

        return result;
    }

    /**
     * Classify anomaly severity
     */
    private classifySeverity(
        anomaly: DetectedAnomaly,
        baseline: BaselineMetrics
    ): 'info' | 'warning' | 'critical' {
        const deviation = Math.abs(anomaly.deviationPercent);

        // Severity based on deviation and confidence
        if (deviation > 50 && anomaly.confidence > 0.8) {
            return 'critical';
        } else if (deviation > 25 && anomaly.confidence > 0.7) {
            return 'warning';
        }

        return 'info';
    }

    /**
     * Identify potential root cause
     */
    private identifyRootCause(
        anomaly: DetectedAnomaly,
        input: AnomalyInput
    ): string | undefined {
        const causes: Record<string, string[]> = {
            velocity: [
                'Team capacity changes',
                'Sprint scope changes',
                'Blocking dependencies',
                'Technical complexity',
            ],
            pattern: [
                'Process deviation',
                'Workflow changes',
                'External factors',
                'Data quality issues',
            ],
            security: [
                'New vulnerability introduced',
                'Configuration drift',
                'Access pattern change',
                'External threat',
            ],
            quality: [
                'Code complexity increase',
                'Test coverage decrease',
                'Deployment issues',
                'Integration problems',
            ],
            resource: [
                'Team availability',
                'Infrastructure capacity',
                'Budget constraints',
                'Skill gaps',
            ],
        };

        const typeCauses = causes[anomaly.type] || causes['pattern'];

        // Simple heuristic: pick based on deviation direction
        if (anomaly.currentValue < anomaly.baselineValue) {
            return typeCauses[0]; // First cause for decrease
        } else {
            return typeCauses[1]; // Second cause for increase
        }
    }

    /**
     * Generate recommendations based on detected anomalies
     */
    private generateRecommendations(
        anomalies: DetectedAnomaly[]
    ): AIRecommendation[] {
        const recommendations: AIRecommendation[] = [];

        const criticalAnomalies = anomalies.filter((a) => a.severity === 'critical');
        const warningAnomalies = anomalies.filter((a) => a.severity === 'warning');

        if (criticalAnomalies.length > 0) {
            recommendations.push({
                id: 'investigate-critical',
                type: 'action',
                title: 'Investigate critical anomalies immediately',
                description: `${criticalAnomalies.length} critical anomalies detected that require immediate attention`,
                priority: 'high',
                action: {
                    type: 'navigate',
                    target: '/ai/alerts',
                    parameters: { filter: 'critical' },
                    impact: 'low',
                    confidence: 0.95,
                    requiresConfirmation: false,
                },
            });
        }

        if (warningAnomalies.length > 2) {
            recommendations.push({
                id: 'review-warnings',
                type: 'review',
                title: 'Review warning anomalies',
                description: 'Multiple warning-level anomalies detected. Consider scheduling a review.',
                priority: 'medium',
            });
        }

        // Type-specific recommendations
        const velocityAnomalies = anomalies.filter((a) => a.type === 'velocity');
        if (velocityAnomalies.length > 0) {
            recommendations.push({
                id: 'velocity-analysis',
                type: 'analysis',
                title: 'Analyze velocity trends',
                description: 'Velocity anomalies detected. Review team capacity and sprint planning.',
                priority: velocityAnomalies.some((a) => a.severity === 'critical') ? 'high' : 'medium',
            });
        }

        return recommendations;
    }

    /**
     * Calculate overall confidence
     */
    private calculateOverallConfidence(anomalies: DetectedAnomaly[]): number {
        if (anomalies.length === 0) {
            return 1.0; // High confidence that no anomalies exist
        }

        const avgConfidence =
            anomalies.reduce((sum, a) => sum + a.confidence, 0) / anomalies.length;
        return Math.round(avgConfidence * 100) / 100;
    }

    /**
     * Check dependencies
     */
    protected async checkDependencies(): Promise<
        Record<string, 'healthy' | 'degraded' | 'unhealthy'>
    > {
        return {
            statisticalEngine: 'healthy',
            isolationForest: 'healthy',
            lstmDetector: 'healthy',
            alertService: this.alertService ? 'healthy' : 'degraded',
        };
    }
}

File: products/yappc/core/yappc-domain/src/main/java/com/ghatana/yappc/domain/agent/AnomalyDetectorAgent.java
package com.ghatana.products.yappc.domain.agent;

import com.ghatana.platform.observability.MetricsCollector;
import io.activej.promise.Promise;
import io.activej.promise.Promises;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Anomaly Detector Agent for real-time anomaly detection.
 * <p>
 * Uses ensemble anomaly detection algorithms to identify unusual patterns
 * in metrics, behaviors, and system states.
 * <p>
 * <b>Responsibilities:</b>
 * <ul>
 *   <li>Real-time anomaly detection in metrics streams</li>
 *   <li>Pattern break identification</li>
 *   <li>Root cause analysis</li>
 *   <li>Alert generation and severity classification</li>
 * </ul>
 * <p>
 * <b>Models:</b> Isolation Forest, LSTM
 * <p>
 * <b>Latency SLA:</b> 200ms
 *
 * @doc.type class
 * @doc.purpose Real-time anomaly detection agent
 * @doc.layer product
 * @doc.pattern AIAgent
 */
public class AnomalyDetectorAgent extends AbstractAIAgent<AnomalyInput, AnomalyOutput> {

    private static final Logger LOG = LoggerFactory.getLogger(AnomalyDetectorAgent.class);

    private static final String VERSION = "2.0.0";
    private static final String DESCRIPTION = "Real-time anomaly detection in metrics and patterns";
    private static final List<String> CAPABILITIES = List.of(
            "anomaly-detection",
            "pattern-analysis",
            "root-cause-analysis",
            "severity-classification",
            "real-time-monitoring"
    );
    private static final List<String> SUPPORTED_MODELS = List.of(
            "isolation-forest",
            "lstm",
            "statistical"
    );

    private final AnomalyMLService mlService;
    private final AlertService alertService;
    private final MetricStore metricStore;

    /**
     * Creates a new AnomalyDetectorAgent.
     *
     * @param mlService        The ML service for anomaly detection
     * @param alertService     The alert service for notifications
     * @param metricStore      The metric store for baseline data
     * @param metricsCollector The metrics collector
     */
    public AnomalyDetectorAgent(
            @NotNull AnomalyMLService mlService,
            @NotNull AlertService alertService,
            @NotNull MetricStore metricStore,
            @NotNull MetricsCollector metricsCollector
    ) {
        super(
                AgentName.ANOMALY_DETECTOR_AGENT,
                VERSION,
                DESCRIPTION,
                CAPABILITIES,
                SUPPORTED_MODELS,
                metricsCollector
        );
        this.mlService = mlService;
        this.alertService = alertService;
        this.metricStore = metricStore;
    }

    @Override
    public void validateInput(@NotNull AnomalyInput input) {
        if (input.metricType() == null) {
            throw new IllegalArgumentException("Metric type is required");
        }
        if (input.currentMetrics() == null || input.currentMetrics().isEmpty()) {
            throw new IllegalArgumentException("Current metrics cannot be empty");
        }
        if (input.sensitivity() < 0 || input.sensitivity() > 1) {
            throw new IllegalArgumentException("Sensitivity must be between 0 and 1");
        }
    }

    @Override
    protected Promise<ProcessResult<AnomalyOutput>> processRequest(
            @NotNull AnomalyInput input,
            @NotNull AIAgentContext context
    ) {
        LOG.debug("Detecting anomalies for metric type: {}", input.metricType());

        // Load baseline metrics
        return metricStore.getBaseline(context.workspaceId(), input.metricType(), "7d")
                .then(baseline -> {
                    // Run anomaly detection with ensemble
                    return runEnsembleDetection(input, baseline, context);
                })
                .then(result -> {
                    // Classify severity and identify root causes
                    List<AnomalyOutput.DetectedAnomaly> classifiedAnomalies =
                            classifyAndEnrichAnomalies(result.anomalies(), result.baseline(), input);

                    // Trigger alerts for high-severity anomalies
                    return triggerAlerts(classifiedAnomalies, context)
                            .map(v -> {
                                // Generate mitigation recommendations
                                List<AnomalyOutput.MitigationStep> recommendations =
                                        generateMitigationSteps(classifiedAnomalies);

                                AnomalyOutput output = AnomalyOutput.builder()
                                        .anomaliesDetected(classifiedAnomalies.size())
                                        .anomalies(classifiedAnomalies)
                                        .baseline(result.baseline())
                                        .recommendations(recommendations)
                                        .build();

                                double avgConfidence = classifiedAnomalies.stream()
                                        .mapToDouble(AnomalyOutput.DetectedAnomaly::confidence)
                                        .average()
                                        .orElse(0.8);

                                return ProcessResult.of(output, avgConfidence);
                            });
                });
    }

    private Promise<DetectionResult> runEnsembleDetection(
            AnomalyInput input,
            AnomalyOutput.Baseline baseline,
            AIAgentContext context
    ) {
        // Run multiple detection algorithms in parallel
        Promise<List<RawAnomaly>> isolationForest = mlService.detectWithIsolationForest(
                input.currentMetrics(),
                baseline,
                input.sensitivity()
        );

        Promise<List<RawAnomaly>> lstm = input.timeSeriesData() != null
                ? mlService.detectWithLSTM(input.timeSeriesData(), 100)
                : Promise.of(List.of());

        Promise<List<RawAnomaly>> statistical = mlService.detectWithStatistical(
                input.currentMetrics(),
                baseline,
                3.0 // 3 sigma threshold
        );

        // Combine results
        return Promises.toList(isolationForest, lstm, statistical)
                .map(results -> {
                    List<RawAnomaly> allAnomalies = new ArrayList<>();
                    for (List<RawAnomaly> result : results) {
                        allAnomalies.addAll(result);
                    }

                    // Ensemble voting - anomalies detected by multiple algorithms have higher confidence
                    List<RawAnomaly> votedAnomalies = ensembleVote(allAnomalies);

                    return new DetectionResult(votedAnomalies, baseline);
                });
    }

    private List<RawAnomaly> ensembleVote(List<RawAnomaly> allAnomalies) {
        // Group anomalies by metric and timestamp
        Map<String, List<RawAnomaly>> grouped = new java.util.HashMap<>();
        for (RawAnomaly anomaly : allAnomalies) {
            String key = anomaly.metricName() + ":" + anomaly.timestamp();
            grouped.computeIfAbsent(key, k -> new ArrayList<>()).add(anomaly);
        }

        // Keep anomalies detected by at least 2 algorithms, boost confidence for unanimous
        List<RawAnomaly> voted = new ArrayList<>();
        for (Map.Entry<String, List<RawAnomaly>> entry : grouped.entrySet()) {
            List<RawAnomaly> group = entry.getValue();
            if (group.size() >= 2) {
                // Average the confidence, boost if detected by all algorithms
                double avgConfidence = group.stream()
                        .mapToDouble(RawAnomaly::confidence)
                        .average()
                        .orElse(0.5);

                if (group.size() == 3) {
                    avgConfidence = Math.min(avgConfidence * 1.2, 1.0);
                }

                RawAnomaly first = group.get(0);
                voted.add(new RawAnomaly(
                        first.metricName(),
                        first.timestamp(),
                        first.value(),
                        first.baselineValue(),
                        avgConfidence,
                        group.stream()
                                .flatMap(a -> a.algorithms().stream())
                                .distinct()
                                .toList()
                ));
            }
        }

        return voted;
    }

    private List<AnomalyOutput.DetectedAnomaly> classifyAndEnrichAnomalies(
            List<RawAnomaly> rawAnomalies,
            AnomalyOutput.Baseline baseline,
            AnomalyInput input
    ) {
        List<AnomalyOutput.DetectedAnomaly> classified = new ArrayList<>();

        for (RawAnomaly raw : rawAnomalies) {
            double deviationPercent = ((raw.value() - raw.baselineValue()) / raw.baselineValue()) * 100;

            // Determine anomaly type based on metric and deviation
            AnomalyOutput.DetectedAnomaly.AnomalyType type = determineAnomalyType(
                    input.metricType(),
                    deviationPercent
            );

            // Classify severity
            AnomalyOutput.DetectedAnomaly.Severity severity =
                    AnomalyOutput.DetectedAnomaly.Severity.fromScore(
                            Math.abs(deviationPercent) / 100 * raw.confidence()
                    );

            // Attempt root cause analysis
            AnomalyOutput.RootCauseAnalysis rootCause = analyzeRootCause(raw, input);

            classified.add(new AnomalyOutput.DetectedAnomaly(
                    "anomaly-" + UUID.randomUUID().toString().substring(0, 8),
                    type,
                    severity,
                    generateAnomalyTitle(type, raw.metricName()),
                    generateAnomalyDescription(type, raw, deviationPercent),
                    Instant.now().toString(),
                    List.of(), // affected items to be filled
                    raw.baselineValue(),
                    raw.value(),
                    deviationPercent,
                    raw.confidence(),
                    VERSION,
                    rootCause
            ));
        }

        return classified;
    }

    private AnomalyOutput.DetectedAnomaly.AnomalyType determineAnomalyType(
            AnomalyInput.AnomalyMetricType metricType,
            double deviationPercent
    ) {
        return switch (metricType) {
            case VELOCITY -> deviationPercent < 0
                    ? AnomalyOutput.DetectedAnomaly.AnomalyType.VELOCITY_DROP
                    : AnomalyOutput.DetectedAnomaly.AnomalyType.VELOCITY_SPIKE;
            case ERROR_RATE, QUALITY_METRICS -> AnomalyOutput.DetectedAnomaly.AnomalyType.QUALITY_DEGRADATION;
            case SECURITY_EVENTS -> AnomalyOutput.DetectedAnomaly.AnomalyType.SECURITY_THREAT;
            case RESOURCE_USAGE -> AnomalyOutput.DetectedAnomaly.AnomalyType.RESOURCE_EXHAUSTION;
            default -> AnomalyOutput.DetectedAnomaly.AnomalyType.PATTERN_BREAK;
        };
    }

    private String generateAnomalyTitle(
            AnomalyOutput.DetectedAnomaly.AnomalyType type,
            String metricName
    ) {
        return switch (type) {
            case VELOCITY_DROP -> "Velocity Drop Detected";
            case VELOCITY_SPIKE -> "Unusual Velocity Spike";
            case PATTERN_BREAK -> "Pattern Deviation in " + metricName;
            case SECURITY_THREAT -> "Security Anomaly Detected";
            case QUALITY_DEGRADATION -> "Quality Metrics Degradation";
            case RESOURCE_EXHAUSTION -> "Resource Usage Anomaly";
            case UNUSUAL_ACCESS -> "Unusual Access Pattern";
            case CUSTOM -> "Anomaly in " + metricName;
        };
    }

    private String generateAnomalyDescription(
            AnomalyOutput.DetectedAnomaly.AnomalyType type,
            RawAnomaly anomaly,
            double deviationPercent
    ) {
        return String.format(
                "%s detected a %.1f%% deviation from baseline. " +
                        "Current value: %.2f, Baseline: %.2f. Detected by: %s",
                type.name().replace("_", " "),
                deviationPercent,
                anomaly.value(),
                anomaly.baselineValue(),
                String.join(", ", anomaly.algorithms())
        );
    }

    private AnomalyOutput.RootCauseAnalysis analyzeRootCause(RawAnomaly anomaly, AnomalyInput input) {
        // Simplified root cause analysis
        List<String> factors = new ArrayList<>();

        if (input.context() != null) {
            if (input.context().containsKey("recentDeployment")) {
                factors.add("Recent deployment detected");
            }
            if (input.context().containsKey("teamChanges")) {
                factors.add("Team composition changes");
            }
        }

        if (factors.isEmpty()) {
            factors.add("External factors under investigation");
        }

        return new AnomalyOutput.RootCauseAnalysis(
                "automated",
                "Preliminary root cause analysis",
                0.6,
                factors
        );
    }

    private Promise<Void> triggerAlerts(
            List<AnomalyOutput.DetectedAnomaly> anomalies,
            AIAgentContext context
    ) {
        List<AnomalyOutput.DetectedAnomaly> highSeverity = anomalies.stream()
                .filter(a -> a.severity() == AnomalyOutput.DetectedAnomaly.Severity.CRITICAL)
                .toList();

        if (highSeverity.isEmpty()) {
            return Promise.complete();
        }

        return alertService.trigger(new AlertRequest(
                "anomaly-detected",
                highSeverity.stream().mapToDouble(AnomalyOutput.DetectedAnomaly::getSeverityScore).max().orElse(0.7),
                highSeverity,
                context.workspaceId(),
                context.userId()
        ));
    }

    private List<AnomalyOutput.MitigationStep> generateMitigationSteps(
            List<AnomalyOutput.DetectedAnomaly> anomalies
    ) {
        List<AnomalyOutput.MitigationStep> steps = new ArrayList<>();

        for (AnomalyOutput.DetectedAnomaly anomaly : anomalies) {
            switch (anomaly.type()) {
                case VELOCITY_DROP -> steps.add(new AnomalyOutput.MitigationStep(
                        "mit-velocity",
                        "Review Team Capacity",
                        "Investigate blockers and resource constraints affecting velocity",
                        AnomalyOutput.MitigationStep.MitigationPriority.HIGH,
                        AnomalyOutput.MitigationStep.MitigationType.IMMEDIATE,
                        Map.of("action", "review-blockers")
                ));
                case SECURITY_THREAT -> steps.add(new AnomalyOutput.MitigationStep(
                        "mit-security",
                        "Security Investigation Required",
                        "Initiate security review and audit recent access patterns",
                        AnomalyOutput.MitigationStep.MitigationPriority.URGENT,
                        AnomalyOutput.MitigationStep.MitigationType.IMMEDIATE,
                        Map.of("action", "security-audit")
                ));
                case QUALITY_DEGRADATION -> steps.add(new AnomalyOutput.MitigationStep(
                        "mit-quality",
                        "Quality Review",
                        "Review recent changes and increase test coverage",
                        AnomalyOutput.MitigationStep.MitigationPriority.MEDIUM,
                        AnomalyOutput.MitigationStep.MitigationType.SHORT_TERM,
                        Map.of("action", "quality-review")
                ));
                default -> steps.add(new AnomalyOutput.MitigationStep(
                        "mit-investigate",
                        "Investigate Anomaly",
                        "Further investigation required to determine root cause",
                        AnomalyOutput.MitigationStep.MitigationPriority.MEDIUM,
                        AnomalyOutput.MitigationStep.MitigationType.SHORT_TERM,
                        Map.of("anomalyId", anomaly.id())
                ));
            }
        }

        return steps;
    }

    @Override
    protected Promise<Map<String, AgentHealth.DependencyStatus>> doHealthCheck() {
        return mlService.healthCheck()
                .map(healthy -> Map.of(
                        "mlService", healthy ? AgentHealth.DependencyStatus.HEALTHY : AgentHealth.DependencyStatus.UNHEALTHY,
                        "alertService", AgentHealth.DependencyStatus.HEALTHY,
                        "metricStore", AgentHealth.DependencyStatus.HEALTHY
                ));
    }

    // Service interfaces and internal types

    private record DetectionResult(List<RawAnomaly> anomalies, AnomalyOutput.Baseline baseline) {}

    private record RawAnomaly(
            String metricName,
            String timestamp,
            double value,
            double baselineValue,
            double confidence,
            List<String> algorithms
    ) {}

    /**
     * ML Service for anomaly detection.
     */
    public interface AnomalyMLService {
        Promise<List<RawAnomaly>> detectWithIsolationForest(
                List<AnomalyInput.MetricDataPoint> metrics,
                AnomalyOutput.Baseline baseline,
                double sensitivity
        );

        Promise<List<RawAnomaly>> detectWithLSTM(
                List<AnomalyInput.MetricDataPoint> timeSeriesData,
                int lookbackWindow
        );

        Promise<List<RawAnomaly>> detectWithStatistical(
                List<AnomalyInput.MetricDataPoint> metrics,
                AnomalyOutput.Baseline baseline,
                double sigmaThreshold
        );

        Promise<Boolean> healthCheck();
    }

    /**
     * Alert service for notifications.
     */
    public interface AlertService {
        Promise<Void> trigger(AlertRequest request);
    }

    /**
     * Metric store for baseline data.
     */
    public interface MetricStore {
        Promise<AnomalyOutput.Baseline> getBaseline(
                String workspaceId,
                AnomalyInput.AnomalyMetricType metricType,
                String timeWindow
        );
    }

    /**
     * Alert request.
     */
    public record AlertRequest(
            String type,
            double severity,
            List<AnomalyOutput.DetectedAnomaly> anomalies,
            String workspaceId,
            String userId
    ) {}
}
